File: ArraysClass.py
class TArrays():
    def __init__(self):
        self.DataMatrix: list[list[float]] = [[]]
        self.SubjectArray: list[int] = []
        self.SubjectIndexArray: list[int] = []
        self.BContrastArray: list[list[float]] = [[]]
        self.WContrastArray: list[list[float]] = [[]]
        
        self.Header: list[str] = []
        self.GroupsList: list[str] = []
        self.BetweenComments: list[str] = []
        self.WithinComments: list[str] = []

        self.SumOfSubjects: int = 0
        self.NumberOfGroups: int = 0
        self.NumberOfRepeats: int = 0
        self.NumberOfBContrasts: int = 0
        self.NumberOfWContrasts: int = 0

        self.AlphaString: str = ""
        self.DoConfidenceInterval: int = 0
        self.DoRescaling: bool = False

        self.CC: list[float] = [0.0, 1.0, 2.0]
        self.Alpha: float = 0.0
        self.p: int = 0
        self.q: int = 0
        self.OrdB: int = 0
        self.OrdW: int = 0
        self.DFE: float = 0
        self.RoyMessage: bool = False

        self.Bcc: float = 0.0
        self.Wcc: float = 0.0
        self.BWcc: float = 0.0

    def clear_all(self):
        self.DataMatrix = [[]]
        self.SubjectArray = []
        self.SubjectIndexArray = []
        self.BContrastArray = [[]]
        self.WContrastArray = [[]]
        
        self.Header = []
        self.GroupsList = []
        self.BetweenComments = []
        self.WithinComments = []

        self.SumOfSubjects = 0
        self.NumberOfGroups = 0
        self.NumberOfRepeats = 0
        self.NumberOfBContrasts = 0
        self.NumberOfWContrasts = 0

        self.AlphaString = ""
        self.DoConfidenceInterval = 0
        self.DoRescaling = False

        self.CC = [0.0, 1.0, 2.0]
        self.Alpha = 0.0
        self.p = 0
        self.q = 0
        self.OrdB = 0
        self.OrdW = 0
        self.DFE = 0
        self.RoyMessage = False

        self.Bcc = 0.0
        self.Wcc = 0.0
        self.BWcc= 0.0
        

File: PSY-source_notes/ArraysClass.pas
unit ArraysClass;

interface

uses
  Defines, Classes;

type
  TArrays = class
  public
    DataMatrix                 : array of array of Float; // group x row x col
    SubjectArray               : array of Integer;
    SubjectIndexArray          : array of Integer;
    // nb use float for these arrays to avoid incorect results when squaring
    //    large coefs, ie 99999
    BContrastArray             : array of array of Float;
    WContrastArray             : array of array of Float;
//    RepeatComments             : TStringList;
//    GroupComments              : TStringList;
    Header                     : TStringList;
    GroupsList                 : TStringList;
    BetweenComments            : TStringList;
    WithinComments             : TStringList;

    SumOfSubjects              : Integer;
    NumberOfGroups             : Integer;
    NumberOfRepeats            : Integer;
    NumberOfBContrasts         : Integer;
    NumberOfWContrasts         : Integer;

//   OutFileName                : string;
//   OutFileTitle               : string;
//   OutFileText                : Text;
//   PrintText                  : Text;
    AlphaString                : string;
//   FullFileName               : string;
//   FileToSave                 : string;

//   AContrastsOrthogonal       : Boolean;
//   ResultsToScreen            : Boolean;
//   OpenForAnalysis            : Boolean;
    DoConfidenceInterval       : Byte;
    DoRescaling                : Boolean;

    CC                         : array [0..2] of Float;
    Alpha                      : Float;
    p, q                       : Integer;
    OrdB, OrdW                 : Integer;
    DFE                        : Float;
    RoyMessage                 : Boolean;

    Bcc                        : Float;          // Between critical constant
    Wcc                        : Float;          // Within critical constant
    BWcc                       : Float;          // B & W ineraction

    procedure ClearAll;
  end;

var
  Arrays: TArrays = nil;

implementation

procedure TArrays.ClearAll;
begin
    SetLength(DataMatrix, 1, 1);
    SetLength(SubjectArray, 1);
    SetLength(SubjectIndexArray, 1);
    SetLength(BContrastArray, 1, 1);
    SetLength(WContrastArray, 1, 1);

    DataMatrix[0,0] := 0.0;
    SubjectArray[0] := 0;
    SubjectIndexArray[0] := 0;
    BContrastArray[0,0] := 0.0;
    WContrastArray[0,0] := 0.0;
    
    SumOfSubjects := 0;
    NumberOfGroups := 0;
    NumberOfRepeats := 0;
    NumberOfBContrasts := 0;
    NumberOfWContrasts := 0;
    AlphaString := '';
    DoConfidenceInterval := 1;
    DoRescaling := False;
    CC[0] := 0.0;
    CC[1] := 0.0;
    CC[2] := 0.0;
    Alpha := 0.0;
    p := 0;
    q := 0;
    OrdB := 0;
    OrdW := 0;
    DFE := 0.0;
    RoyMessage := False;
    Bcc := 0.0;
    Wcc := 0.0;
    BWcc := 0.0;

    Header.Clear;
    GroupsList.Clear;
    BetweenComments.Clear;
    WithinComments.Clear;
end;

{ initialise and finalise here (globally) }
initialization
  Arrays := TArrays.Create;
  with Arrays do
  begin
    Header := TStringList.Create;
    GroupsList := TStringList.Create;
    BetweenComments := TStringList.Create;
    WithinComments := TStringList.Create;
  end;

finalization
  with Arrays do
  begin
    Header.Free;
    GroupsList.Free;
    BetweenComments.Free;
    WithinComments.Free;
  end;
  Arrays.Free;

end.
